crates/burn-autodiff/src/runtime/memory_management.rs

```diff
@@ -115,81 +107,98 @@ impl GraphMemoryManagement {
-                let node_status = self
-                    .statuses
-                    .get(&node_id)
-                    .expect("All nodes should have received a status at this point")
-                    .clone();
-
-                match node_status {
-                    NodeMemoryStatus::Useful => {
-                        // Nothing to do, was already tagged through some other path
-                    }
-                    NodeMemoryStatus::Unavailable => {
-                        // Even if this node is unavailable, it is still possible that an ancestor is useful if referenced
-                        for parent in parents {
-                            self.useful_propagation(parent, Mode::Explore);
-                        }
-                    }
-                    NodeMemoryStatus::Unknown => {
-                        // If this node is referenced and not unavailable,
-                        // then it is useful and we must retain all ancestors
-
-                        let mut mode = Mode::Explore;
-                        if self.is_referenced(node_id) {
-                            self.statuses.insert(node_id, NodeMemoryStatus::Useful);
-                            mode = Mode::TagAsUseful;
-                        }
 
-                        for parent in parents {
-                            self.useful_propagation(parent, mode.clone());
-                        }
-                    }
+            for parent in self
+                .nodes
+                .get(&node_id)
+                .cloned()
+                .unwrap_or_default()
+                .into_iter()
+            {
+                if !visited_as_useful.contains(&parent)
+                    && (Some(&NodeMemoryStatus::Useful) == self.statuses.get(&node_id)
+                        || !visited_as_unknown.contains(&parent))
+                {
+                    to_visit.push((parent, Some(node_id)));
                 }
             }
         }
     }
 
-    fn is_referenced(&self, node_id: NodeID) -> bool {
-        match self.nodes.get_key_value(&node_id) {
-            Some((key, _value)) => Arc::strong_count(key) > 1,
-            None => panic!("Node should be in the nodes map"),
-        }
-    }
-
     fn identify_leaves_and_deletables(
         &self,
-        node_id: NodeID,
+        leaf_id: NodeID,
         new_leaves: &mut HashSet<NodeID>,
         to_delete: &mut Vec<NodeID>,
     ) {
-        let current_status = self
-            .statuses
-            .get(&node_id)
-            .expect("Node should have status");
-
-        match current_status {
-            NodeMemoryStatus::Useful => {
-                new_leaves.insert(node_id);
-            }
-            _ => {
-                let parents = self.nodes.get(&node_id).cloned().unwrap_or(vec![]);
-                for parent in parents {
-                    self.identify_leaves_and_deletables(parent, new_leaves, to_delete)
+        let mut visited = HashSet::new();
```

Comment:
Should be initialized with the right capacity, which is `new_leaves.len()`

Reply:
Hmm `new_leaves` is dynamically inserted during this algorithm, it's length 0 at that moment. 
Also we can't really know `visited` length in advance because it depends if the nodes the algorithm sees are useful.